code/how-to-spawn-vim/src/main.rs

```diff
@@ -0,0 +1,87 @@
+// ANCHOR: all
+// ANCHOR: imports
+use ratatui::{
+    backend::CrosstermBackend,
+    crossterm::{
+        event::{self, KeyCode, KeyEventKind},
+        terminal::{disable_raw_mode, enable_raw_mode, EnterAlternateScreen, LeaveAlternateScreen},
+        ExecutableCommand,
+    },
+    widgets::Paragraph,
+    Terminal,
+};
+use std::io::{stdout, Result};
+use std::process::Command;
+// ANCHOR_END: imports
+
+// ANCHOR: action_enum
+enum Action {
+    Quit,
+    EditFile,
+}
+// ANCHOR_END: action_enum
+
+// ANCHOR: main
+fn main() -> Result<()> {
+    // ANCHOR: setup
+    stdout().execute(EnterAlternateScreen)?;
+    enable_raw_mode()?;
+    let mut terminal = Terminal::new(CrosstermBackend::new(stdout()))?;
+    terminal.clear()?;
+    // ANCHOR_END: setup
+
+    loop {
+        // ANCHOR: draw
+        terminal.draw(|frame| {
+            let area = frame.size();
+            frame.render_widget(
+                Paragraph::new("Hello ratatui! (press 'q' to quit, 'e' to edit a file)"),
+                area,
+            );
+        })?;
+        // ANCHOR_END: draw
+
+        // ANCHOR: handle-events
+        // Check if there's any event
+        if event::poll(std::time::Duration::from_millis(16))? {
+            // Read the event
+            if let event::Event::Key(key) = event::read()? {
+                if key.kind == KeyEventKind::Press {
+                    let action = match key.code {
+                        KeyCode::Char('q') => Some(Action::Quit),
+                        KeyCode::Char('e') => Some(Action::EditFile),
+                        _ => None,
+                    };
+
+                    if let Some(action) = action {
+                        match action {
+                            Action::Quit => break,
+                            // ANCHOR: action-handler_EditFile
+                            Action::EditFile => {
+                                stdout().execute(LeaveAlternateScreen)?;
+                                disable_raw_mode()?;
+                                // Launch vim and wait for status
+                                // You may change this to other editors like nvim, nano, etc.
+                                Command::new("vim").arg("/tmp/a.txt").status()?;
+                                stdout().execute(EnterAlternateScreen)?;
+                                enable_raw_mode()?;
+                                // re-init tui
+                                terminal = ratatui::Terminal::new(CrosstermBackend::new(stdout()))?;
```

Comment:
```suggestion
                                terminal.clear()?;
```

I just tested this and it's recovering the terminal just fine. I also tested the same logic on component template and it's working. I think this is what you were looking for @kdheepak from our discord conversation. Here are new changes for component template that spawns vim and restores terminal correctly without using `tui = tui::Tui::new()?.tick_rate(self.tick_rate).frame_rate(self.frame_rate);

```diff
diff --git a/src/action.rs b/src/action.rs
index f422a62..ee157c2 100644
--- a/src/action.rs
+++ b/src/action.rs
@@ -17,4 +17,5 @@ pub enum Action {
   Refresh,
   Error(String),
   Help,
+  EditFile,
 }
diff --git a/src/app.rs b/src/app.rs
index e528d05..e01a35a 100644
--- a/src/app.rs
+++ b/src/app.rs
@@ -105,6 +105,7 @@ impl App {
           Action::Quit => self.should_quit = true,
           Action::Suspend => self.should_suspend = true,
           Action::Resume => self.should_suspend = false,
+          Action::Refresh => tui.terminal.clear()?,
           Action::Resize(w, h) => {
             tui.resize(Rect::new(0, 0, w, h))?;
             tui.draw(|f| {
@@ -126,6 +127,18 @@ impl App {
               }
             })?;
           },
+          Action::EditFile => {
+            tui.exit()?;
+            let status = std::process::Command::new("vim").arg("/tmp/a.txt").status()?;
+            if status.success() {
+              log::info!("Successfully launched vim");
+            } else {
+              log::error!("Failed to launch vim");
+            }
+            // tui = tui::Tui::new()?.tick_rate(self.tick_rate).frame_rate(self.frame_rate);
+            action_tx.send(Action::Refresh)?;
+            tui.enter()?;
+          },
           _ => {},
         }
         for component in self.components.iter_mut() {
@@ -137,7 +150,8 @@ impl App {
       if self.should_suspend {
         tui.suspend()?;
         action_tx.send(Action::Resume)?;
-        tui = tui::Tui::new()?.tick_rate(self.tick_rate).frame_rate(self.frame_rate);
+        // tui = tui::Tui::new()?.tick_rate(self.tick_rate).frame_rate(self.frame_rate);
+        action_tx.send(Action::Refresh)?;
         // tui.mouse(true);
         tui.enter()?;
       } else if self.should_quit {
diff --git a/src/components/home.rs b/src/components/home.rs
index a1e9cc6..7426597 100644
--- a/src/components/home.rs
+++ b/src/components/home.rs
@@ -35,10 +35,19 @@ impl Component for Home {
     Ok(())
   }
 
+  fn handle_key_events(&mut self, key: KeyEvent) -> Result<Option<Action>> {
+    match key.code {
+      KeyCode::Char('q') => Ok(Some(Action::Quit)),
+      KeyCode::Char('r') => Ok(Some(Action::Refresh)),
+      KeyCode::Char('v') => Ok(Some(Action::EditFile)),
+      _ => Ok(None),
+    }
+  }
+
   fn update(&mut self, action: Action) -> Result<Option<Action>> {
     match action {
-      Action::Tick => {
-      },
+      Action::Render => {},
+      Action::Tick => {},
       _ => {},
     }
     Ok(None)
```

Here's the definition of `terminal.clear()` in crossterm:
```rust
    /// Clear the terminal and force a full redraw on the next draw call.
    pub fn clear(&mut self) -> io::Result<()> {
        match self.viewport {
            Viewport::Fullscreen => self.backend.clear_region(ClearType::All)?,
            Viewport::Inline(_) => {
                self.backend
                    .set_cursor(self.viewport_area.left(), self.viewport_area.top())?;
                self.backend.clear_region(ClearType::AfterCursor)?;
            }
            Viewport::Fixed(area) => {
                for row in area.top()..area.bottom() {
                    self.backend.set_cursor(0, row)?;
                    self.backend.clear_region(ClearType::AfterCursor)?;
                }
            }
        }
        // Reset the back buffer to make sure the next update will redraw everything.
        self.buffers[1 - self.current].reset();
        Ok(())
    }
```

Reply:
Yeah that works!

Reply:
Should I create a PR on templates repo to add this minor `Action::Refresh` update so that it's fixed there?

Reply:
Yes please! That'd be awesome!

Reply:
Let's name it `Action::ClearScreen`? To make it easier to understand. I don't need we need a `Action::Refresh`.

Reply:
Sure, will do :+1:

Reply:
Here's the pr with these changes on templates repo: https://github.com/ratatui-org/templates/pull/61